python/python/bystro/supervised_ppca/gf_dropout_pt.py

```diff
@@ -0,0 +1,299 @@
+
+        if task == "classification":
+            sigm = nn.Sigmoid()
+            supervision_loss = nn.BCELoss()
+        elif task == "regression":
+            supervision_loss = nn.MSELoss()
+        else:
+            supervision_loss = nn.BCELoss()
+
+        trainable_variables = [W_, sigmal_, B_]
+
+        optimizer = torch.optim.SGD(
+            trainable_variables,
+            lr=training_options["learning_rate"],
+            momentum=training_options["momentum"],
+        )
+        eye = torch.tensor(np.eye(p))
+        one_s = torch.tensor(np.ones(self.n_supervised))
+        softplus = nn.Softplus()
+
+        _prior = self._create_prior()
+
+        myrange = trange if progress_bar else range
+
+        for i in myrange(training_options["n_iterations"]):
+            idx = rng.choice(
+                X.shape[0], size=training_options["batch_size"], replace=False
+            )
+            X_batch = X[idx]
+            y_batch = y[idx]
+
+            sigma = softplus(sigmal_)
+            WWT = torch.matmul(torch.transpose(W_, 0, 1), W_)
+            Sigma = WWT + sigma * eye
+
+            like_prior = _prior(trainable_variables)
+
+            # Generative likelihood
+            m = MultivariateNormal(torch.zeros(p), Sigma)
+            like_gen = torch.mean(m.log_prob(X_batch))
+
+            # Predictive lower bound
+            P_x, Cov = _get_projection_matrix(W_, sigma)
+            mean_z = torch.matmul(X_batch, torch.transpose(P_x, 0, 1))
+            eps = torch.rand_like(mean_z)
+            C1_2 = torch.cholesky(Cov)
+            z_samples = mean_z + torch.matmul(eps, C1_2)
+
+            y_hat = (
+                self.delta
+                * torch.matmul(z_samples[:, : self.n_supervised], one_s)
+                + B_
+            )
+            if task == "regression":
+                loss_y = supervision_loss(y_hat, y_batch)
+            else:
+                loss_y = supervision_loss(sigm(y_hat), y_batch)
+
+            WTW = torch.matmul(W_, torch.transpose(W_, 0, 1))
+            off_diag = WTW - torch.diag(torch.diag(WTW))
+            torch.norm(off_diag)
+
+            posterior = like_gen + 1 / N * like_prior
+            loss = -1 * posterior + self.mu * loss_y
+
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            self._save_losses(i, like_gen, like_prior, posterior)
+            self.losses_supervision[i] = loss_y.detach().numpy()
+
+        self._store_instance_variables(trainable_variables)
+        return self
+
+    def _store_instance_variables(self, trainable_variables):
+        """
+        Saves the learned variables
+
+        Parameters
+        ----------
+        trainable_variables : list
+            List of tensorflow variables saved
+
+        Sets
+        ----
+        W_ : np.array-like,(n_components,p)
+            The loadings
+
+        sigmas_ : np.array-like,(n_components,p)
```

Comment:
No or at least they aren't the same. Sigmal_ is the latent unconstrained variable we optimize with respect to. Sigma2_ is the isotropic variance of PPCA while sigmas_ is the not isotropic diagonal in factor analysis.

Reply:
Maybe I can try to rephrase my question:

The docstring for this method says that it sets three variables, `W_`, `sigmas_` and `B_`.  But the method actually sets `W_`, `sigma2_` and `B_`.  

If there's a discrepancy between what the docstring says the method does and what the method does, we should resolve this discrepancy so that future developers aren't confused.

On the other hand, if the docstring and method are in fact correct as written (i.e. the docstring should say we set `sigmas_` and the method should set `sigma2_`), that's probably worth a comment explaining the distinction.

AIUI this class--because it implements PPCA--shouldn't have even have a `sigmas_` attribute, right?

Reply:
Yeah that's correct. Are we planning on making docstrings standard in all modules in Bystro?

Reply:
> Yeah that's correct. Are we planning on making docstrings standard in all modules in Bystro?

Yeah, as you and I chatted offline, for public methods, I think we should, since we'll be auto-generating documentation off of it.